datalad/cmd.py

```diff
@@ -141,27 +143,106 @@ def run_gitcommand_on_file_list_chunks(func, cmd, files, *args, **kwargs):
     return ''.join(out), ''.join(err)
 
 
-def kill_output(output):
-    """Helper for WitlessRunner to swallow all output and neither
-    relay it to the parent process's stdout, nor provide it as the
-    return value of WitlessRunner.run().
-    """
-    return b'', 0
+async def run_async_cmd(loop, cmd, protocol, stdin, **kwargs):
+    lgr.debug('Async run %s', cmd)
+
+    cmd_done = asyncio.Future(loop=loop)
+    factory = functools.partial(protocol, cmd_done)
+    proc = loop.subprocess_exec(
+        factory,
+        *cmd,
+        stdin=stdin,
+        # ask the protocol which streams to capture
+        stdout=asyncio.subprocess.PIPE if protocol.proc_out else None,
+        stderr=asyncio.subprocess.PIPE if protocol.proc_err else None,
+        **kwargs
+    )
+    transport = None
+    try:
+        lgr.debug('Launching process %s', cmd)
+        transport, protocol = await proc
+        lgr.debug('Waiting for process %i to complete', transport.get_pid())
+        await cmd_done
+    finally:
+        # protect against a crash whe launching the process
+        if transport:
+            transport.close()
+
+    return cmd_done.result()
+
+
+class WitlessProtocol(asyncio.SubprocessProtocol):
+
+    FD_NAMES = ['stdin', 'stdout', 'stderr']
+
+    proc_out = None
+    proc_err = None
+
+    def __init__(self, done_future):
+        # future promise to be fulfilled when process exits
+        self.done = done_future
+        # capture output in bytearrays while the process is running
+        Streams = namedtuple('Streams', ['out', 'err'])
+        self.buffer = Streams(
+            out=bytearray() if self.proc_out else None,
+            err=bytearray() if self.proc_err else None,
+        )
+        self.pid = None
+        super().__init__()
 
+    def connection_made(self, transport):
+        self.transport = transport
+        self.pid = transport.get_pid()
+        lgr.debug('Process %i started', self.pid)
+
+    def pipe_data_received(self, fd, data):
+        lgr.debug('Read %i bytes from %i[%s]', len(data), self.pid, self.FD_NAMES[fd])
```

Comment:
I think it's probably reasonable to assume this could be called many times and that `data` could be somewhat large.  In that case, it probably makes sense to guard this `.debug` call with `logger.isEnabledFor`.  (The same applies to `KillOutput.pipe_data_received`.)

Reply:
True!

Reply:
Done.